doorman/jwt_generic.go

```diff
@@ -0,0 +1,189 @@
+package doorman
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/allegro/bigcache"
+	jose "gopkg.in/square/go-jose.v2"
+	jwt "gopkg.in/square/go-jose.v2/jwt"
+)
+
+// OpenIDConfiguration is the OpenID provider metadata about endpoints etc.
+type OpenIDConfiguration struct {
+	JWKSUri string `json:"jwks_uri"`
+}
+
+// JWKS are the JWT public keys
+type JWKS struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+type jwtGenericValidator struct {
+	Issuer             string
+	ClaimExtractor     ClaimExtractor
+	SignatureAlgorithm jose.SignatureAlgorithm
+	cache              *bigcache.BigCache
+}
+
+// newJWTGenericValidator returns a generic JWT validator of this issuer.
+func newJWTGenericValidator(issuer string, extractor ClaimExtractor) *jwtGenericValidator {
+	cache, _ := bigcache.NewBigCache(bigcache.DefaultConfig(1 * time.Hour))
+
+	if extractor == nil {
+		extractor = &defaultClaimExtractor{}
+	}
+
+	return &jwtGenericValidator{
+		Issuer:             issuer,
+		ClaimExtractor:     extractor,
+		SignatureAlgorithm: jose.RS256,
+		cache:              cache,
+	}
+}
+
+func (v *jwtGenericValidator) config() (*OpenIDConfiguration, error) {
+	cacheKey := "config:" + v.Issuer
+	data, err := v.cache.Get(cacheKey)
+
+	// Cache is empty or expired: fetch again.
+	if err != nil {
+		uri := strings.TrimRight(v.Issuer, "/") + "/.well-known/openid-configuration"
+		log.Debugf("Fetch OpenID configuration from %s", uri)
+		data, err = downloadJSON(uri)
+		if err != nil {
+			return nil, err
```

Comment:
look into this [errors](https://godoc.org/github.com/pkg/errors) package. So this would what the code would look like: `return nil, errors.Wrap(err, "Failed downloading issuer")`.  One of the issues with propagating errors up the call stack in go is that you lose context of where the error occurred.  A network error three function calls deep gives no context. 

However, if you can "unwrap" your errors starting from: "Failed downloading issuer" it is a lot easier to debug where the error came from. This [blog post](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) explains it in detail a lot better. Worthy read.